src/encoded/ingestion/loadxl_extensions.py

```diff
@@ -93,29 +100,93 @@ def package_loadxl_response(loadxl_response: Generator[bytes, None, None]) -> Di
         )
         return response
 
-    def call_loadxl(validate_only: bool):
-        nonlocal portal_vapp, data, post_only, patch_only
-        return package_loadxl_response(
-                loadxl_load_data(
-                    testapp=portal_vapp,
-                    inserts=data,
-                    docsdir=None,
-                    overwrite=True,
-                    itype=None,
-                    from_json=True,
-                    continue_on_exception=True,
-                    verbose=True,
-                    post_only=post_only,
-                    patch_only=patch_only,
-                    validate_only=validate_only,
-                    skip_links=True))
-
-    if validate_first and not validate_only:
-        response = call_loadxl(validate_only=True)
-        if response.get("errors", []):
-            return response
-
-    return call_loadxl(validate_only=validate_only)
+    def define_progress_tracker(submission_uuid: str, validation: bool, total: int,
+                                vapp: Optional[VirtualApp] = None) -> Optional[Callable]:
+        nonlocal ingestion_status
+        progress_status_datetime_values = [PROGRESS_LOADXL.START,
+                                           PROGRESS_LOADXL.START_SECOND_ROUND,
+                                           PROGRESS_LOADXL.DONE]
+        progress_status_string_values = [PROGRESS_LOADXL.MESSAGE,
+                                         PROGRESS_LOADXL.MESSAGE_VERBOSE,
+                                         PROGRESS_LOADXL.MESSAGE_DEBUG]
+        progress_status = {}
+        for progress_status_enum in PROGRESS_LOADXL.values():
+            if ((progress_status_enum in progress_status_datetime_values) or
+                (progress_status_enum in progress_status_string_values)):  # noqa
+                progress_status[progress_status_enum] = None
+            else:
+                progress_status[progress_status_enum] = 0
+        progress_status = {**progress_status,
+                           PROGRESS_LOADXL.TOTAL: total,
+                           PROGRESS_INGESTER.VALIDATION: validation}
+        def progress_tracker(progress: PROGRESS_LOADXL) -> None:  # noqa
+            nonlocal progress_status
+            def progress_message() -> None:  # noqa
+                # Just a convenience/courtesy so the consumer (smaht-submitr) doesn't have to cobble
+                # together a status message; but the data is still there of course if they want/need to.
+                nonlocal progress_status, total, validate_only
```

Comment:
Is use of `nonlocal` necessary here? This entire block above as well is a little hard to follow